README.rst
===================
SIDESPLITTER plugin
===================

This plugin provides a wrapper for `SIDESPLITTER <https://github.com/StructuralBiology-ICLMedicine/SIDESPLITTER>`_ program for mitigating local overfitting.

.. image:: https://img.shields.io/pypi/v/scipion-em-sidesplitter.svg
        :target: https://pypi.python.org/pypi/scipion-em-sidesplitter
        :alt: PyPI release

.. image:: https://img.shields.io/pypi/l/scipion-em-sidesplitter.svg
        :target: https://pypi.python.org/pypi/scipion-em-sidesplitter
        :alt: License

.. image:: https://img.shields.io/pypi/pyversions/scipion-em-sidesplitter.svg
        :target: https://pypi.python.org/pypi/scipion-em-sidesplitter
        :alt: Supported Python versions

.. image:: https://img.shields.io/sonar/quality_gate/scipion-em_scipion-em-sidesplitter?server=https%3A%2F%2Fsonarcloud.io
        :target: https://sonarcloud.io/dashboard?id=scipion-em_scipion-em-sidesplitter
        :alt: SonarCloud quality gate

.. image:: https://img.shields.io/pypi/dm/scipion-em-sidesplitter
        :target: https://pypi.python.org/pypi/scipion-em-sidesplitter
        :alt: Downloads

Installation
------------

You will need to use 3.0+ version of Scipion to be able to run these protocols. To install the plugin, you have two options:

a) Stable version

.. code-block::

    scipion installp -p scipion-em-sidesplitter

b) Developer's version

    * download repository

    .. code-block::

        git clone -b devel https://github.com/scipion-em/scipion-em-sidesplitter.git

    * install

    .. code-block::

        scipion installp -p /path/to/scipion-em-sidesplitter --devel

SIDESPLITTER sources will be installed automatically with the plugin, but you can also link an existing installation.

    * Default installation path assumed is ``software/em/sidesplitter-1.2``, if you want to change it, set *SIDESPLITTER_HOME* in ``scipion.conf`` file pointing to the folder where the SIDESPLITTER is installed.

To check the installation, simply run one of the following Scipion tests:

.. code-block::

   scipion test sidesplitter.tests.test_protocol_sidesplitter.TestSideSplitter

A complete list of tests can also be seen by executing ``scipion test --show --grep sidesplitter``

Supported versions
------------------

1.2

Protocols
---------

* local filter

References
----------

1. Kailash Ramlaul, Colin M. Palmer and Christopher H. S. Aylett (2020). Mitigating Local Over-fitting During Single Particle Reconstruction with SIDESPLITTER. https://www.biorxiv.org/content/10.1101/2019.12.12.874081v2


CHANGES.txt
3.0.12 - maintenance release
3.0.11 - add possible outputs dict, update installation procedure
3.0.10 - minor changes
3.0.9 - fix output filenames
3.0.7 - fix compilation flags, add devStatus
3.0.6 - fix failing test
3.0.4 - fix installation and compilation
3.0.3 - fix installation, update to 1.2
3.0.2 - changed version to reflect Scipion 3 support
1.0.2 - remove unused code
1.0.0 - First plugin version


sidesplitter/tests/__init__.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from .test_protocol_sidesplitter import ProtSideSplitter


sidesplitter/tests/test_protocol_sidesplitter.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import os

from pyworkflow.utils import magentaStr, makePath, copyFile
from pyworkflow.tests import BaseTest, DataSet, setupTestProject
from pyworkflow.plugin import Domain
from pwem.objects import Volume
from pwem.protocols import ProtImportParticles, ProtImportVolumes

from ..protocols import ProtSideSplitter


try:
    ProtRelionCreateMask3D = Domain.importFromPlugin("relion.protocols",
                                                     "ProtRelionCreateMask3D",
                                                     doRaise=True)
    ProtRelionRefine3D = Domain.importFromPlugin("relion.protocols",
                                                 "ProtRelionRefine3D",
                                                 doRaise=True)
except ImportError as e:
    print("Relion plugin not found! You need to install it to be able to run this test.")


class TestSideSplitter(BaseTest):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        cls.ds = DataSet.getDataSet('relion_tutorial')
        pathFns = 'import/refine3d/extra'
        cls.volFn = cls.ds.getFile(os.path.join(pathFns, 'relion_class001.mrc'))
        cls.half1Fn = cls.ds.getFile(os.path.join(pathFns, 'relion_it025_half1_class001.mrc'))
        cls.half2Fn = cls.ds.getFile(os.path.join(pathFns, 'relion_it025_half2_class001.mrc'))

    def importVolume(self):
        print(magentaStr("\n==> Importing data - volume:"))
        protVol = self.newProtocol(ProtImportVolumes,
                                   objLabel='import volume',
                                   filesPath=self.volFn,
                                   samplingRate=3)
        self.launchProtocol(protVol)
        return protVol

    def importPartsFromScipion(self):
        print(magentaStr("\n==> Importing data - particles:"))
        partFn = self.ds.getFile('import/particles.sqlite')
        protPart = self.newProtocol(ProtImportParticles,
                                    objLabel='Import Particles',
                                    importFrom=ProtImportParticles.IMPORT_FROM_SCIPION,
                                    sqliteFile=partFn,
                                    magnification=10000,
                                    samplingRate=3,
                                    haveDataBeenPhaseFlipped=True
                                    )
        self.launchProtocol(protPart)
        return protPart

    def _createRef3DProtBox(self, label, protocol):
        from pyworkflow.protocol.constants import STATUS_FINISHED

        prot = self.newProtocol(protocol)
        self.saveProtocol(prot)

        prot.setObjLabel(label)
        makePath(prot._getPath())
        makePath(prot._getExtraPath())
        makePath(prot._getTmpPath())

        prot.inputParticles.set(self.importPartsFromScipion().outputParticles)

        outputVol = self.importVolume().outputVolume
        prot.referenceVolume.set(outputVol)

        volume = Volume()
        volume.setFileName(prot._getExtraPath('test.mrc'))
        pxSize = prot.inputParticles.get().getSamplingRate()
        volume.setSamplingRate(pxSize)

        prot._defineOutputs(outputVolume=volume)
        prot.setStatus(STATUS_FINISHED)

        # Create a mask protocol
        print(magentaStr("\n==> Running relion - create mask 3d:"))
        protMask = self.newProtocol(ProtRelionCreateMask3D)
        protMask.inputVolume.set(outputVol)
        self.launchProtocol(protMask)

        return prot, protMask

    def _validations(self, vol, dims, pxSize):
        self.assertIsNotNone(vol, "There was a problem with sidesplitter "
                                  "protocol, using Relion auto-refine as input")
        xDim = vol.getXDim()
        sr = vol.getSamplingRate()
        self.assertEqual(xDim, dims, "The dimension of your volume is (%d)^3 "
                                     "and must be (%d)^3" % (xDim, dims))

        self.assertAlmostEqual(sr, pxSize, delta=0.0001,
                               msg="Pixel size of your volume is %0.2f and"
                                   " must be %0.2f" % (sr, pxSize))

    def test_sidesplitter(self):
        protRef, protMask = self._createRef3DProtBox("auto-refine",
                                                     ProtRelionRefine3D)
        protRef._createFilenameTemplates()
        volPath = protRef._getFileName('finalvolume', ref3d=1).split(':')[0]
        volHalf1 = protRef._getFileName('final_half1_volume', ref3d=1).split(':')[0]
        volHalf2 = protRef._getFileName('final_half2_volume', ref3d=1).split(':')[0]

        copyFile(self.volFn, volPath)
        copyFile(self.half1Fn, volHalf1)
        copyFile(self.half2Fn, volHalf2)

        protRef.outputVolume.setFileName(volPath)
        protRef.outputVolume.setHalfMaps([volHalf1, volHalf2])
        project = protRef.getProject()
        project._storeProtocol(protRef)

        print(magentaStr("\n==> Testing sidesplitter - after refine 3d:"))
        sidesplitterProt = self.newProtocol(ProtSideSplitter,
                                            protRefine=protRef,
                                            mask=protMask.outputMask)
        sidesplitterProt.setObjLabel('sidesplitter after Auto-refine')

        self.launchProtocol(sidesplitterProt)
        self._validations(sidesplitterProt.outputVolume1, 60, 3)


sidesplitter/bibtex.py
# coding: latin-1
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************
"""
@article{Ramlaul2020,
title = "Mitigating Local Over-fitting During Single Particle Reconstruction with SIDESPLITTER",
journal = "BioRxiv",
year = "2020",
doi = "https://dx.doi.org/10.1101/2019.12.12.874081",
url = "https://www.biorxiv.org/content/10.1101/2019.12.12.874081v2",
author = "Kailash Ramlaul, Colin M. Palmer and Christopher H. S. Aylett",
}

"""

sidesplitter/convert.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from pwem.emlib.image import ImageHandler


def getImageLocation(location):
    return ImageHandler.locationToXmipp(location)


def convertMask(img, outFn, newDim=None):
    """ Convert binary mask to a format read by Relion and truncate the
    values between 0-1 values, due to Relion only support masks with this
    values (0-1).
    Params:
        img: input image to be converted.
        outFn: output file path.
    Return:
        new file name of the mask.
    """

    ih = ImageHandler()
    imgFn = getImageLocation(img.getLocation())
    ih.truncateMask(imgFn, outFn, newDim=newDim)

    return outFn


sidesplitter/__init__.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import os

import pwem
from pyworkflow.utils import Environ

from .constants import SIDESPLITTER_HOME, V1_2

__version__ = '3.0.12'
_logo = "sidesplitter_logo.png"
_references = ['Ramlaul2020']


class Plugin(pwem.Plugin):
    _homeVar = SIDESPLITTER_HOME
    _pathVars = [SIDESPLITTER_HOME]
    _supportedVersions = V1_2
    _url = "https://github.com/scipion-em/scipion-em-sidesplitter"

    @classmethod
    def _defineVariables(cls):
        cls._defineEmVar(SIDESPLITTER_HOME, 'sidesplitter-1.2')

    @classmethod
    def getEnviron(cls):
        """ Setup the environment variables needed to launch sidesplitter. """
        environ = Environ(os.environ)
        environ.update({'PATH': cls.getHome()}, position=Environ.BEGIN)
        return environ

    @classmethod
    def getProgram(cls):
        """ Return the program binary that will be used. """
        cmd = cls.getHome('sidesplitter')
        return str(cmd)

    @classmethod
    def defineBinaries(cls, env):
        ver = "1.2"
        url = 'https://github.com/StructuralBiology-ICLMedicine/SIDESPLITTER.git'
        installCmd = [
            f'cd .. && rmdir sidesplitter-{ver} &&',
            f'git clone {url} sidesplitter-{ver} &&',
            f'cd sidesplitter-{ver} &&',
            'gcc -O3 *.c -lm -pthread -lfftw3 -lfftw3_threads -std=c99 -o sidesplitter'
        ]

        commands = [(" ".join(installCmd), 'sidesplitter')]

        env.addPackage('sidesplitter', version=ver,
                       tar='void.tgz',
                       neededProgs=['git', 'gcc'],
                       commands=commands,
                       default=True)


sidesplitter/constants.py
# **************************************************************************
# *
# *  Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************


SIDESPLITTER_HOME = 'SIDESPLITTER_HOME'

# Supported versions
V1_2 = '1.2'


sidesplitter/protocols/protocol_sidesplitter.py
# **************************************************************************
# *
# * Authors:     Grigory Sharov (gsharov@mrc-lmb.cam.ac.uk)
# *
# * MRC Laboratory of Molecular Biology (MRC-LMB)
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 3 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import os

import pyworkflow.protocol.params as params
from pyworkflow.constants import PROD
from pwem.protocols import ProtAnalysis3D
from pwem.objects import Volume
from pwem.emlib.image import ImageHandler

from sidesplitter import Plugin
from ..convert import convertMask


class ProtSideSplitter(ProtAnalysis3D):
    """
    Protocol for mitigating local over-fitting by filtering.
    """
    _label = 'local filter'
    _devStatus = PROD
    _possibleOutputs = {
        'outputVolume1': Volume,
        'outputVolume2': Volume
    }

    def _createFilenameTemplates(self):
        """ Centralize how files are called. """
        myDict = {'half1': self._getExtraPath("half1_unfil.mrc"),
                  'half2': self._getExtraPath("half2_unfil.mrc"),
                  'mask': self._getExtraPath("mask.mrc"),
                  'outHalf1Fn': self._getExtraPath('half1_unfil_sidesplitter.mrc'),
                  'outHalf2Fn': self._getExtraPath('half2_unfil_sidesplitter.mrc'),
                  }

        self._updateFilenamesDict(myDict)

    # --------------------------- DEFINE param functions ----------------------

    def _defineParams(self, form):
        form.addSection(label='Input')
        form.addParam('protRefine', params.PointerParam,
                      important=True,
                      pointerClass="ProtRefine3D",
                      label='Select a previous refinement protocol',
                      help='Select any previous refinement protocol to get the '
                           '3D half maps. Note that the refinement protocol '
                           'must use gold-standard method.')
        form.addParam('mask', params.PointerParam,
                      allowsNull=True,
                      pointerClass="VolumeMask",
                      label='Volume mask',
                      help="Provide the mask used in 3D refinement.")
        form.addParam('doSNRWeighting', params.BooleanParam,
                      expertLevel=params.LEVEL_ADVANCED,
                      default=False,
                      label='Use SNR-weighted spectrum',
                      help='Outputs the SNR weighted spectrum rather '
                           'than matching input spectrum / grey-scale.')

        form.addParallelSection(threads=1, mpi=0)

    # --------------------------- INSERT steps functions ----------------------
    
    def _insertAllSteps(self):
        self._createFilenameTemplates()
        self._insertFunctionStep('convertInputStep')
        self._insertFunctionStep('runSideSplitterStep')
        self._insertFunctionStep('createOutputStep')

    # --------------------------- STEPS functions -----------------------------
    
    def convertInputStep(self):
        """ Convert input half-maps to mrc as expected by SIDESPLITTER."""
        protRef = self.protRefine.get()
        outVol = protRef.outputVolume
        dim = outVol.getXDim()
        vols = outVol.getHalfMaps().split(',')
        ih = ImageHandler()

        if self.mask.hasValue():
            convertMask(self.mask.get(), self._getFileName('mask'), newDim=dim)

        for vol, key in zip(vols, ['half1', 'half2']):
            ih.convert(vol, self._getFileName(key))

    def runSideSplitterStep(self):
        """ Call SIDESPLITTER with the appropriate parameters. """
        args = self._getArgs()
        param = ' '.join(['%s %s' % (k, str(v)) for k, v in args.items()])
        program = Plugin.getProgram()
        cmd = 'export OMP_NUM_THREADS=%d; ' % self.numberOfThreads.get()
        cmd += program

        self.runJob(cmd, param, env=Plugin.getEnviron(),
                    cwd=self._getExtraPath(),
                    numberOfThreads=1)

    def createOutputStep(self):
        inputVol = self.protRefine.get().outputVolume
        ps = inputVol.getSamplingRate()

        vol = Volume()
        vol.setSamplingRate(ps)
        vol.setObjLabel('Filtered half-map 1')
        vol.setFileName(self._getFileName('outHalf1Fn'))

        vol2 = Volume()
        vol2.setSamplingRate(ps)
        vol2.setObjLabel('Filtered half-map 2')
        vol2.setFileName(self._getFileName('outHalf2Fn'))

        outputs = {'outputVolume1': vol,
                   'outputVolume2': vol2}
        self._defineOutputs(**outputs)
        self._defineSourceRelation(inputVol, vol)
        self._defineSourceRelation(inputVol, vol2)

    # --------------------------- INFO functions ------------------------------
    
    def _summary(self):
        summary = []

        if hasattr(self, 'outputVolume1'):
            summary.append("Created locally filtered half-maps.")
        else:
            summary.append("Output is not ready")

        return summary
    
    def _validate(self):
        errors = []

        return errors
    
    # --------------------------- UTILS functions -----------------------------
 
    def _getArgs(self):
        """ Prepare the args dictionary."""
        args = {'--v1': os.path.basename(self._getFileName('half1')),
                '--v2': os.path.basename(self._getFileName('half2'))}

        if self.mask.hasValue():
            args['--mask'] = os.path.basename(self._getFileName('mask'))

        if self.doSNRWeighting:
            args['--spectrum'] = ' '

        return args
